design/0002-vscode-extension-wot-cli.org

```diff
@@ -0,0 +1,39 @@
+#+TITLE: VSCode Extension and wot CLI
+#+AUTHOR: Pedram Hadjian
+
+* Goal
+
+The following is a suggestion of a first milestone to be discussed in the community. The ediTDor project currently helps in the creation and validation of Thing Models and Thing Descriptions. It uses JSON Schema to validate the content and adds a visualization of the TM/TD as a rendered interactive HTML site.
+
+There have been many requests to turn this project into a VSCode extension and make the rendering available separately like swagger-ui, to be rendered in code forges like gitlab or github. Also the functionality of validation should be usable in CI/CD pipelines, so it should be available as a separate library/cli.
+
+In addition, while developers can create syntactically valid TMs/TDs, there is no way of knowing, if we can actually use these descriptions to communicate with the modeled device, let alone automatically validate the communication in a CI/CD pipeline.
+
+* Proposal
+
+In short: start to create a VSCode extension, which uses JSON Schemas to validate TMs/TDs. Also provide a cli and a library, which provides the same JSON Schema validation for use in a CI/CD pipeline. In addition the cli should integrate in a simple way with a protocol driver to test communication with a device. To "prove" successful communication, the recorded timeseries should be persistable to a file to be committed to a repository, where this communication can be validated.
+
+I see the following tasks to achieve the above:
+
+- Create a VSCode extension that syntactically validates TMs/TDs
+- Context extensions of JSON-LD may be used to map to JSON Schemas
+- Provide wot-ui to create HTML pages for TMs/TDs
+- Introduce a wot-cli
+- wot-cli can validate TMs/TDs against JSON Schemas
+- wot-cli can generate stubs for a protocol driver to communicate with it
+- wot-cli can be used to communicate with an implementation of those stubs
+- wot-cli can observe properties and events and record them in timeseries
+- wot-cli can generate appropriate CSV protocol bindings
+- wot-cli can add protocol bindings to TMs to exchange field level bindings with CSV ones
```

Comment:
I would align wot-cli with [@thingwebplayground/cli](https://github.com/thingweb/thingweb-playground/tree/master/packages/cli). As that is already doing some of the functionality listed here, @edbingo is also working on extensions. It might make sense to keep editdor as an editor, and leave cli tasks to thingwebplayground.

Reply:
Thanks for your feedback @fennibay. I got an introduction to the existing thingweb project by @egekorkan and I agree with your points. I'll think about how to document these insights.

Reply:
@fennibay we'll go into detail about what task is fulfilled by which component in the scenarios we will document. I agree that https://github.com/thingweb/thingweb-playground/tree/master/packages/cli should be used. The ediTDor uses this project already for validation and this is how the tools should work together. I think there is room for improvement in the way it is used.